lib/streamlit/elements/widgets/radio.py

```diff
@@ -51,46 +53,89 @@
 @dataclass
 class RadioSerde(Generic[T]):
     options: Sequence[T]
-    index: int
+    index: int | None
+
+    def serialize(self, v: object) -> int | None:
+        if v is None:
+            return None
 
-    def serialize(self, v: object) -> int:
-        if len(self.options) == 0:
-            return 0
-        return index_(self.options, v)
+        return 0 if len(self.options) == 0 else index_(self.options, v)
 
     def deserialize(
         self,
-        ui_value: Optional[int],
+        ui_value: int | None,
         widget_id: str = "",
-    ) -> Optional[T]:
+    ) -> T | None:
         idx = ui_value if ui_value is not None else self.index
 
         return (
             self.options[idx]
-            if len(self.options) > 0 and self.options[idx] is not None
+            if idx is not None
+            and len(self.options) > 0
+            and self.options[idx] is not None
             else None
         )
 
 
 class RadioMixin:
-    @gather_metrics("radio")
+    @overload
     def radio(
         self,
         label: str,
         options: OptionSequence[T],
         index: int = 0,
         format_func: Callable[[Any], Any] = str,
-        key: Optional[Key] = None,
-        help: Optional[str] = None,
-        on_change: Optional[WidgetCallback] = None,
-        args: Optional[WidgetArgs] = None,
-        kwargs: Optional[WidgetKwargs] = None,
+        key: Key | None = None,
+        help: str | None = None,
+        on_change: WidgetCallback | None = None,
+        args: WidgetArgs | None = None,
+        kwargs: WidgetKwargs | None = None,
+        *,  # keyword-only args:
+        disabled: bool = False,
+        horizontal: bool = False,
+        captions: Sequence[str] | None = None,
+        label_visibility: LabelVisibility = "visible",
+    ) -> T:
+        pass
+
+    @overload
+    def radio(
+        self,
+        label: str,
+        options: OptionSequence[T],
+        index: None = None,
+        format_func: Callable[[Any], Any] = str,
+        key: Key | None = None,
+        help: str | None = None,
+        on_change: WidgetCallback | None = None,
+        args: WidgetArgs | None = None,
+        kwargs: WidgetKwargs | None = None,
+        *,  # keyword-only args:
+        disabled: bool = False,
+        horizontal: bool = False,
+        captions: Sequence[str] | None = None,
+        label_visibility: LabelVisibility = "visible",
+    ) -> T | None:
+        pass
```

Comment:
Are these duplicates necessary?

Reply:
Not strictly, but this signature overload improves typing for users since radio cannot return `None` if index was not `None`. It's not a big problem for radio, since radio was already annotated with `Optional`. But for some of the other widgets, some users may run into typing issues if we just add `None` to the return types without overloading.

Reply:
Not sure where I land on whether its worth it, but could you add a comment if you choose to leave it?

Reply:
I decided to remove the overloads for radio and selectbox since they cannot fully cover the cases anyways (since empty list also returns `None` as mentioned by Vincent below).